src/plugins/elasticsearch_ui_shared/public/request.ts

```diff
@@ -0,0 +1,151 @@
+}
+
+export interface SendRequestResponse {
+  data: any;
+  error: Error;
+}
+
+export interface UseRequest extends SendRequest {
+  interval?: number;
+  initialData?: any;
+  processData?: any;
+}
+
+export const sendRequest = async (
+  httpClient: any,
+  { path, method, body }: SendRequest
+): Promise<Partial<SendRequestResponse>> => {
+  try {
+    // NOTE: This is tightly coupled to Angular's $http service.
+    const response = await httpClient[method](path, body);
+
+    if (typeof response.data === 'undefined') {
+      throw new Error(response.statusText);
+    }
+
+    return { data: response.data };
+  } catch (e) {
+    return {
+      error: e.response ? e.response : e,
+    };
+  }
+};
+
+export const useRequest = (
+  httpClient: any,
+  { path, method, body, interval, initialData, processData }: UseRequest
+) => {
+  // Main states for tracking request status and data
+  const [error, setError] = useState<null | any>(null);
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+  const [data, setData] = useState<any>(initialData);
+
+  // Consumers can use isInitialRequest to implement a polling UX.
+  const [isInitialRequest, setIsInitialRequest] = useState<boolean>(true);
+  const requestInterval = useRef<any>(null);
+  const requestIntervalId = useRef<any>(null);
+
+  // We always want to use the most recently-set interval in updateInterval.
+  requestInterval.current = interval;
+
+  // Tied to every render and bound to each request.
+  let isOutdatedRequest = false;
+
+  const updateInterval = () => {
+    // Clear current interval
+    if (requestIntervalId.current) {
+      clearTimeout(requestIntervalId.current);
+    }
+
+    // Set new interval
+    if (requestInterval.current) {
+      requestIntervalId.current = setTimeout(_sendRequest, requestInterval.current);
+    }
+  };
+
+  const _sendRequest = async () => {
+    // We don't clear error or data, so it's up to the consumer to decide whether to display the
+    // "old" error/data or loading state when a new request is in-flight.
+    setIsLoading(true);
+
+    const requestBody = {
+      path,
+      method,
+      body,
+    };
+
+    const response = await sendRequest(httpClient, requestBody);
+    let { data: responseData } = response;
+    const { error: responseError } = response;
+
+    if (processData) {
+      responseData = processData(responseData);
+    }
+
+    // If an outdated request has resolved, DON'T update state, but DO allow the processData handler
+    // to execute side effects like update telemetry.
+    if (isOutdatedRequest) {
+      return;
+    }
+
+    setError(responseError);
+    setData(responseData);
+    setIsLoading(false);
+    setIsInitialRequest(false);
+
+    // If we're on an interval, we need to schedule the next request. This also allows us to reset
+    // the interval if the user has manually requested the data, to avoid doubled-up requests.
+    updateInterval();
+  };
+
+  useEffect(() => {
```

Comment:
I am not sure I follow the use case of watching for a changes of path or method, as those are separate requests in the UI.

Reply:
I think this is the part that actually performs the request (calling `_sendRequest` outside of `useEffect` will call it on every re-render)

would be good to have a comment to clarify this, though I think just watching on `path` instead of both `path` and `method` is sufficient

Reply:
Ah I see! What if we simplified it and not provide anything in the array? It would be clearer that we want to execute the `_sendRequest` once. I don't see a use case where we would update the path or the method in our component as we use a separate instance of useRequest for each API endpoint.

Reply:
There's nothing preventing someone from using `useRequest` directly inside of a component, and because this is a hook I was striving for functional correctness. To me, that means that if a hook's input changes then its output will reflect the change in input. In a perfect world that would mean changing the request if method, path, or body changes but hey, it's not a perfect world so 🤷‍♂ . I'm going to remove the `method` since I do think this an unlikely thing to change, but I'm going to leave `path` for now since I can imagine the changing and I'll update the comment to explain my reasoning here.